lib/parse.js

```diff
@@ -8,33 +8,73 @@ var defaults = {
     allowDots: false,
     allowPrototypes: false,
     arrayLimit: 20,
+    charset: 'utf-8',
     decoder: utils.decode,
     delimiter: '&',
     depth: 5,
+    interpretNumericEntities: false,
     parameterLimit: 1000,
     plainObjects: false,
-    strictNullHandling: false
+    strictNullHandling: false,
+    utf8Sentinel: false
 };
 
+var interpretNumericEntities = function (str) {
+    return str.replace(/&#(\d+);/g, function ($0, numberStr) {
+        return String.fromCharCode(parseInt(numberStr, 10));
+    });
+};
+
+// This is what browsers will submit when the ✓ character occurs in an
+// application/x-www-form-urlencoded body and the encoding of the page containing
+// the form is iso-8859-1, or when the submitted form has an accept-charset
+// attribute of iso-8859-1. Presumably also with other charsets that do not contain
+// the ✓ character, such as us-ascii.
+var isoSentinel = 'utf8=' + encodeURIComponent('&#10003;'); // %26%2310003%3B
+
+// These are the percent-encoded utf-8 octets representing a checkmark, indicating
+// that the request actually is utf-8 encoded.
+var utf8Sentinel = 'utf8=' + encodeURIComponent('✓'); // %E2%9C%93
+
 var parseValues = function parseQueryStringValues(str, options) {
     var obj = {};
     var cleanStr = options.ignoreQueryPrefix ? str.replace(/^\?/, '') : str;
     var limit = options.parameterLimit === Infinity ? undefined : options.parameterLimit;
     var parts = cleanStr.split(options.delimiter, limit);
+    var charset = options.charset;
+    var i;
+
+    if (options.utf8Sentinel) {
+        for (i = 0; i < parts.length; ++i) {
+            if (parts[i].indexOf('utf8=') === 0) {
+                if (parts[i] === utf8Sentinel) {
```

Comment:
Then we'd need to introduce a `foundSentinel` variable that gets set in each of the three cases so that we can break out of the loop? Or put the `skipIndex=i; i=parts.length;` bit in all three cases?

Reply:
... We could also define that a `utf8=...` parameter we don't explicitly recognize as a checkmark should just be parsed as a regular parameter? I'm in doubt myself.